### ISSUE TITLE
gopls diagnostics clear linter/Problems output

### ISSUE THREAD

**Issue by kb-sp** [golang/vscode-go#3511]
### What version of Go, VS Code & VS Code Go extension are you using?
<details><summary>Version Information</summary><br>

* Run `go version` to get version of Go from _the VS Code integrated terminal_.
	- `go version go1.23.0 darwin/arm64`
* Run `gopls -v version` to get version of Gopls from _the VS Code integrated terminal_.
```
Build info
----------
golang.org/x/tools/gopls v0.16.1
    golang.org/x/tools/gopls@v0.16.1 h1:1hO/dCeUvjEYx3V0rVvCtOkwnpEpqS29paE+Jw4dcAc=
    github.com/BurntSushi/toml@v1.2.1 h1:9F2/+DoOYIOksmaJFPw1tGFy1eDnIJXg+UHjuD8lTak=
    github.com/google/go-cmp@v0.6.0 h1:ofyhxvXcZhMsU5ulbFiLKl/XBFqE1GSq7atu8tAmTRI=
    golang.org/x/exp/typeparams@v0.0.0-20221212164502-fae10dda9338 h1:2O2DON6y3XMJiQRAS1UWU+54aec2uopH3x7MAiqGW6Y=
    golang.org/x/mod@v0.18.0 h1:5+9lSbEzPSdWkH32vYPBwEpX8KwDbM52Ud9xBUvNlb0=
    golang.org/x/sync@v0.7.0 h1:YsImfSBoP9QPYL0xyKJPq0gcaJdG3rInoqxTWbfQu9M=
    golang.org/x/telemetry@v0.0.0-20240607193123-221703e18637 h1:3Wt8mZlbFwG8llny+t18kh7AXxyWePFycXMuVdHxnyM=
    golang.org/x/text@v0.16.0 h1:a94ExnEXNtEwYLGJSIUxnWoxoRz/ZcCsV63ROupILh4=
    golang.org/x/tools@v0.22.1-0.20240628205440-9c895dd76b34 h1:Kd+Z5Pm6uwYx3T2KEkeHMHUMZxDPb/q6b1m+zEcy62c=
    golang.org/x/vuln@v1.0.4 h1:SP0mPeg2PmGCu03V+61EcQiOjmpri2XijexKdzv8Z1I=
    honnef.co/go/tools@v0.4.7 h1:9MDAWxMoSnB6QoSqiVr7P5mtkT9pOc1kSxchzPCnqJs=
    mvdan.cc/gofumpt@v0.6.0 h1:G3QvahNDmpD+Aek/bNOLrFR2XC6ZAdo62dZu65gmwGo=
    mvdan.cc/xurls/v2@v2.5.0 h1:lyBNOm8Wo71UknhUs4QTFUNNMyxy2JEIaKKo0RWOh+8=
go: go1.23.0
```
* Run `code -v` or `code-insiders -v` to get version of VS Code or VS Code Insiders.
```
1.92.2
fee1edb8d6d72a0ddff41e5f71a671c23ed924b9
arm64
```
* Check your installed extensions to get the version of the VS Code Go extension
	- `v0.42.0`
* Run Ctrl+Shift+P (Cmd+Shift+P on Mac OS) > `Go: Locate Configured Go Tools` command.
```
# Tools Configuration


## Environment

GOBIN: undefined
toolsGopath: 
gopath: /Users/me/go
GOROOT: /Users/me/local/Cellar/go/1.23.0/libexec
PATH: /Users/me/google-cloud-sdk/bin:/opt/X11/bin:/opt/local/sbin:/opt/local/bin:/opt/sbin:/opt/bin:/usr/local/sbin:/usr/local/bin:/usr/X11/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Library/Apple/usr/bin

## Tools

	go:	/Users/me/local/bin/go: go version go1.23.0 darwin/arm64

	gopls:	/Users/me/go/bin/gopls	(version: v0.16.1 built with go: go1.23.0)
	gotests:	/Users/me/go/bin/gotests	(version: v1.6.0 built with go: go1.23.0)
	gomodifytags:	/Users/me/go/bin/gomodifytags	(version: v1.16.0 built with go: go1.23.0)
	impl:	/Users/me/go/bin/impl	(version: v1.1.0 built with go: go1.23.0)
	goplay:	/Users/me/go/bin/goplay	(version: v1.0.0 built with go: go1.23.0)
	dlv:	/Users/me/go/bin/dlv	(version: v1.23.0 built with go: go1.23.0)
	golangci-lint:	/Users/me/go/bin/golangci-lint	(version: v1.60.3 built with go: go1.23.0)

## Go env

Workspace Folder (madhatter): /Users/me/go/src/github.com/MyCo/madhatter

	GO111MODULE=''
	GOARCH='arm64'
	GOBIN=''
	GOCACHE='/Users/me/Library/Caches/go-build'
	GOENV='/Users/me/Library/Application Support/go/env'
	GOEXE=''
	GOEXPERIMENT=''
	GOFLAGS=''
	GOHOSTARCH='arm64'
	GOHOSTOS='darwin'
	GOINSECURE=''
	GOMODCACHE='/Users/me/go/pkg/mod'
	GONOPROXY='github.com/MyCo,buf.build/gen/go,github.com/github-handle'
	GONOSUMDB='github.com/MyCo,buf.build/gen/go,github.com/github-handle'
	GOOS='darwin'
	GOPATH='/Users/me/go'
	GOPRIVATE='github.com/MyCo,buf.build/gen/go,github.com/github-handle'
	GOPROXY='https://proxy.golang.org,direct'
	GOROOT='/Users/me/local/Cellar/go/1.23.0/libexec'
	GOSUMDB='sum.golang.org'
	GOTMPDIR=''
	GOTOOLCHAIN='auto'
	GOTOOLDIR='/Users/me/local/Cellar/go/1.23.0/libexec/pkg/tool/darwin_arm64'
	GOVCS=''
	GOVERSION='go1.23.0'
	GODEBUG=''
	GOTELEMETRY='local'
	GOTELEMETRYDIR='/Users/me/Library/Application Support/go/telemetry'
	GCCGO='gccgo'
	GOARM64='v8.0'
	AR='ar'
	CC='cc'
	CXX='c++'
	CGO_ENABLED='0'
	GOMOD='/Users/me/go/src/github.com/MyCo/madhatter/go.mod'
	GOWORK='/Users/me/go/src/github.com/MyCo/madhatter/go.work'
	CGO_CFLAGS='-O2 -g'
	CGO_CPPFLAGS=''
	CGO_CXXFLAGS='-O2 -g'
	CGO_FFLAGS='-O2 -g'
	CGO_LDFLAGS='-O2 -g'
	PKG_CONFIG='pkg-config'
	GOGCCFLAGS='-fPIC -arch arm64 -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -ffile-prefix-map=/var/folders/9z/mn1s9hm917b11k64wnvjm7nh0000gn/T/go-build2463239726=/tmp/go-build -gno-record-gcc-switches -fno-common'
```
</details>

### Share the Go related settings you have added/edited
```
  "go.lintTool": "golangci-lint",
  "go.lintOnSave": "workspace",
```

### Describe the bug

When saving a Go file, `golangci-lint` runs and shows warnings in the Problems pane. But when `gopls` diagnostics are applied, the Problems list is completely cleared of all Go issues. Other linter issues (e.g., for JSON) remain in the Problems pane.

The OUTPUT pane shows the successful run of the linter and its output, and PROBLEMS shows the errors. But when `gopls.diagnosticsDelay` is hit, the PROBLEMS pane is erased, and no additional output appears in the OUTPUT pane.

REALIZATION: Only `golangci-lint` output from the `staticcheck` linter seems to be affected. Other linter issues for Go or other languages remain.

Even if the `staticcheck` linter has become deprecated, this behavior is *super* confusing. ;) Additionally, our CI/CD is gated by a standalone `golangci-lint` run, so it's more than cosmetic.

### Steps to reproduce the behavior:

Create a file that fails the `staticcheck` linter (in `golangci-lint`), save it, and `gopls` diagnostics will quickly clear these entries.

`.golangci.yaml`:
```
linters:
  enable:
    staticcheck
```

**Comment by findleyr** [comment]
@hyangah this sounds like a bug: linters from multiple sources should be merged, no?

> REALIZATION: Only golangci-lint output from the staticcheck linter seems to be affected. Other linter issues for Go or other languages remain.

This might because all of those other linters (for Go files) are actually being run by gopls. Gopls runs the standard vet suite, as well as staticcheck if configured with `"staticcheck": true`.

**Comment by kb-sp** [comment]
I'm not aware of having that on (it's not in our config, and I can't find a Settings entry that has it enabled (or configurable). If it's `gopls.staticcheck` in settings.json, making it `false` doesn't change the behavior. However, marking it manually to `true` _does_ retain the messages, but they're in the broader, less-configurable context of gopls/vet, and doesn't match our golangci-lint config gating our CI/CD. Perhaps `gopls.staticcheck=false` shouldn't _remove_ existing `staticcheck` linters?

**Comment by hyangah** [comment]
The extension is supposed to merge and dedup the diagnostics. When deduping, it favors the reports from gopls. Can you share an example code with problems that exhibit the issue?

**Comment by kb-sp** [comment]
I originally made an initial pass at a test case, but a simple one-/two-package case in the same module wasn't sufficient. If I could get a pointer to the location of the code that handles the deduping, I could take a look and either run a local build and debug or figure out a potential test case.

**Comment by hyangah** [comment]
Thanks.
The relevant code locations:

* Where gopls diagnostics triggers dedupping:

https://github.com/golang/vscode-go/blob/c7d88dcc06ff98d6ee649163ba2aeee235356694/extension/src/language/goLanguageServer.ts#L602

* Where linter diagnostics are saved:

https://github.com/golang/vscode-go/blob/c7d88dcc06ff98d6ee649163ba2aeee235356694/extension/src/goLint.ts#L44

* Looks like deduping is primarily based on the diagnostic report's start line but I could be wrong:

https://github.com/golang/vscode-go/blob/c7d88dcc06ff98d6ee649163ba2aeee235356694/extension/src/util.ts#L705

Let us know what you find!

**Comment by h9jiang** [comment]
This issue has been open for almost a year. If you could reproduce this issue, we will definitely fix it. 

Closing.

**Comment by alexbozhenko** [comment]
@h9jiang 
Can we please reopen this? I have a minimal reproducer.

<details>

```
mkdir vscodego-3511 && cd vscodego-3511

cat > go.mod <<'EOF'
module repro

go 1.25
EOF

cat > main.go <<'EOF'
package main

import (
      "fmt"
      "os"
)

func Save(v interface{}, path string) { os.WriteFile(path, []byte(fmt.Sprint(v)), 0o600) }

func main() {}
EOF
```

</details>


For this code, we can see that gopls emits two "hint" diagnostics on line 8:
```
# gopls check -severity=hint main.go
main.go:8:13-24: interface{} can be replaced by any
main.go:8:60-81: Replace []byte(fmt.Sprint...) with fmt.Append
```

As I reported in https://github.com/microsoft/vscode/issues/270930, those reports from [modernize](https://pkg.go.dev/golang.org/x/tools/go/analysis/passes/modernize) are "hint" level diagnostics and are not shown under the problems tab, only as a little squiggle under the code:

<img width="1286" height="1142" alt="Image" src="https://github.com/user-attachments/assets/fadab9fa-4cd7-4b1d-95b9-9faa1e3ae3e9" />

However, golangci-lint(v2.12.2) with default config reports the following on the same line:
```
golangci-lint run ./... --no-config
main.go:8:53: Error return value of `os.WriteFile` is not checked (errcheck)
func Save(v interface{}, path string) { os.WriteFile(path, []byte(fmt.Sprint(v)), 0o600) }
                                                    ^
1 issues:
* errcheck: 1
```
I would expect to see that report in VSCode, under the "problems" tab as well.

Only after fixing both cosmetic issues reported by gopls can we see the report from golangci-lint under the "problems" tab:

<img width="1175" height="1067" alt="Image" src="https://github.com/user-attachments/assets/3c65f83d-beea-4a05-9bc5-ce909099ef8a" />


> The extension is supposed to merge and dedup the diagnostics. When deduping, it favors the reports from gopls.

The dedup happens on per-line basis. So, in this example, we have a purely cosmetic report from gopls hiding actual bug report from golangci-lint on the same line. Should dedup be more sophisticated?

**Comment by h9jiang** [comment]
Ack, let me take a look then.

**Comment by cubelitblade** [comment]
Hi @h9jiang @alexbozhenko,

I ran into the same problem and while trying to fix it I found the possible cause, it was like a side effect of the design.

I've adapted the above reproduction example a bit.

```go
package main

import (
	"fmt"
	"os"
)

func Save(v interface{}, path string) { os.WriteFile(path, []byte(fmt.Sprint(v)), 0o600) }

func main() {
	a := 42
	if 42 == a {
		a = 45
	}
}

```

 `os.WriteFile(...) `  triggers golangci-lint-v2's errcheck rule,
`42 == a` triggers the ST1017 rule (Yoda condtion) in staticcheck for both gopls and golangci-lint-v2.
`a = 45` triggers golangci-lint's ineffassign rule.

In the current version, errcheck triggered on line 8 and staticcheck from golangci-lint-v2 triggered on line 13 appear briefly and then disappear when lint is triggered. Only staticcheck from gopls and ineffassign from golangci-lint-v2 remain.

<img width="1280" height="764" alt="Image" src="https://github.com/user-attachments/assets/138ed245-5992-4b19-8be4-95dda9907927" />

I found that this was caused by the current handleDiagnostics deduplication in goLanguageServer.ts. When gopls published the diagnosis results, it called the deduplication logic and cleared the problems reported by the linter.

https://github.com/golang/vscode-go/blob/a3b23f16ac4afc2c0303d08b2915c604c2822ed5/extension/src/language/goLanguageServer.ts#L699-L702

While deDupeDiagnostics in utils.ts only deduplicates by line number, when gopls and other sources report diagnosis results for the same line of code, gopls will take control, completely discard problems from other sources. This is why only ineffassign is retained.

https://github.com/golang/vscode-go/blob/a3b23f16ac4afc2c0303d08b2915c604c2822ed5/extension/src/util.ts#L709-L710

I tried to use the line number + problem code combination to accurately remove duplicates, but then found another difficulty: gopls and golangci-lint-v2 report problems in different formats.

gopls:
| Code            | Message                   | File    | Source |
|-----------------|---------------------------|---------|--------|
| ST1017(default) | don't use Yoda conditions | main.go | ST1017 |

golangci-lint-v2:
|  Code | Message                                         | File    | Source               |
|-------|-------------------------------------------------|---------|---------------------|
| -    | ST1017: don't use Yoda conditions (staticcheck) | main.go | go-golangci-lint-v2 |

The code for the problems reported by golangci-lint-v2 exists only in Messgae, and the format of the reports is different for each linter. For example, the previous code doesn't exist for most linters. 

This creates a dilemma: the design was clearly designed to prevent the same problem from being reported multiple times (e.g., ST1017), but it caused all other problems in the same row to be masked; Accurately removing duplicates also seems difficult and relies on unreliable string comparisons. If we wanted to preserve the full diagnosis, it would probably be easiest to remove the deduplication logic, but that's completely at odds with the original design.

I wonder if there is a better way to solve this problem?

**Comment by alexbozhenko** [comment]
Maybe playing with the level of diagnostic messages from golangci-ling vs gopls can help?
E.g. when deduping, we can prefer the message with the highest severity, so at least something will show up under the "problems" tab?

**Comment by cubelitblade** [comment]
> Maybe playing with the level of diagnostic messages from golangci-ling vs gopls can help? E.g. when deduping, we can prefer the message with the highest severity, so at least something will show up under the "problems" tab?

Thank you for the suggestion to consider severity levels! I just submitted a PR inspired by this idea.

Since gopls diagnostics are always preserved, the PR changes the logic to only drop an external linter’s diagnostic if the gopls diagnostic has an equal or higher severity. This ensures that a high-severity error (like an errcheck Error) will no longer be silently dropped just because gopls has a lower-severity warning on the same line.

While same-severity conflicts (e.g., both are Warnings) will still be deduplicated to keep the UI clean, this prevents the most critical errors from being masked.

**Comment by h9jiang** [comment]
Hi cubelitblade@, 

Thank you for your contribution and the previous discussion. I did take a look at the current diagnostic handling and I find it is really quite hard to understand, so I have done some refactoring.

Based on the discussion, I think we should consider taking severity into consideration. In addition, I want to also consider the start position. (the start line and column).

There are three differences between how gopls and linters are reporting the diagnostics:
- the gopls report things through the LSP which is a stream. Every request contains diags for a specific file. But linter may report diags for the entire package or entire module.
- the gopls report diags following a range (start and end), the linter report normally a position (start only). And the extension will read the file and trying to come up with the end position. (so the end position is "best-effort", not promised to be accurate)
- the diagnostics reported by gopls is managed automatically by the language client lib, the diagnostics reported by the linter is managed separately and manually by the extension.

However, I think the start position reported by the linter should be accurate (I trust golangci-lint consider its popularity).

Hence, here is what I think we can do: When gopls report diagnostics, review the linter's diagnostics. Delete the linter's diag if the linter's diagnostic is in the same position (line and column) and the severity is lower than gopls's.

When linter report diagnostics, review the gopls's diagnostics, present the diag if there is no gopls diagnostics at that position or the gopls diagnostics at that position is lower severity.

Basically what you have described but include start position in to consideration. I may take over your CL for faster iteration (because I made some refactoring and I will share more in screenshots over the upcoming days, if the behavior looks reasonable, I can release a pre-release version v0.57.1 in mid Aug so you guys can give it a try).

On a side note, 
- I think go lint, go build should really be deprecated since they are already covered by the gopls.
- For better experience, I suggest you either disable staticcheck from gopls or from golangci-lint v2. gopls setting: `"ui.diagnostic.staticcheck": false,`. (Though I think it is more efficient to run it within gopls consider gopls have all the caching in place).

Thank you.

**Comment by h9jiang** [comment]
After CL 812360: vscode-go will preserve diags reported by linter as long as the severity is higher than gopls's diagnostics in that line with regression test in place.

<img width="1224" height="357" alt="Image" src="https://github.com/user-attachments/assets/1c0f8e35-79f4-462d-8d2f-60ae166ad9b4" />

I also plan to de-duplicate the diags based on the start position. However, there is a chance it will be noisy. So I will try to release two insider version, one version that include the logic to de-duplicate based on severity and another version that include the logic to de-duplicate based on severity and start pos.

If you ever find it too noisy, come back to this issue and tell me. If I do not receive further feedback regarding the noisiness, I may release the next stable version as is. (with both severity and start pos).

**Comment by h9jiang** [comment]
I just released vscode-go v0.57.2, feel free to give it a try. To enable this extension, you need to find the Go extension and clock "Switch to Prerelease version" in the screenshot below.

<img width="759" height="160" alt="Image" src="https://github.com/user-attachments/assets/f403850e-5286-4022-b09f-8c90288c8c30" />

Once switched, you can find the version information in the right side panel indicating it is v0.57.2

<img width="966" height="326" alt="Image" src="https://github.com/user-attachments/assets/9c8dd0b3-8568-4d80-aaa7-c33ca0367576" />

If the marketplace has outdated data, you can also try the Gear icon and "Install specific version"

<img width="559" height="321" alt="Image" src="https://github.com/user-attachments/assets/171dac0d-895c-47fe-b187-221098b9ce11" />


If the diagnostics become too noisy, you can try to install v0.57.1 (consolidation based on severity only) and feel free to come back to this issue and comment.
